### ISSUE TITLE
Subsequent writes are lost

### ISSUE THREAD

**Issue by ivan-kleshnin** [Yomguithereal/baobab#190]
Even with "chaining":

```javascript
// ACTIONS

// This is caching wrapper for data fetching
function loadMany(page, query) {
  // State.select("robots", "page").set(page); // lost
  // State.select("robots").set("page", page); // lost
  State.select("robots", "page").chain(() => parseInt(page)); // this write is lost as well
  // ... cache access will be here
  return fetchMany(page, query);
}

// This is real data-fetching action
function fetchMany() {
  ...
  return Axios.get(apiURL, {params})
    .then(response => {
      ...
      State.select("robots").merge({ // merge shouldn't influence "page" but it does &_&
        loadError: false,
        total: meta.page && meta.page.total || Object.keys(models).length,
        models: models,
      });
      ...
}
```

As you can see even while first and second writes have zero overlapping keys, data from first write is lost. Perhaps chaining in both cases could help but writes occur in different functions and I can't bridge such dependencies. So despite so many syntax choices I can't make the simplest case working. 
I think this should be considered as an unexpected / bad behavior. Data shouldn't be lost so easy.
Well, data shouldn't be lost at all. Can this chaining be the default behavior except maybe some really tricky cases?

**Comment by Yomguithereal** [comment]
I don't why I did this but I considered `merge` to be a destructive update while it's not and can be combined. So this is indeed a bug with the `merge` command I should fix.

**Comment by ivan-kleshnin** [comment]
Thinking more about it I'm coming to conclusion that the way Baobab behaves now is fundamentally wrong. The good news is that it probably can be fixed. Let me argument my point of view.

One of the major mistakes of React component architecture is how it's state is treated.
`this.state` always refer to rendered state, not pending state. Therefore `setState` is async, accepting a callback. People asked alot about `setState` returning a promise but React devs were negative about this. There is a symmentry between `setState` and `render` with second accepting a callback which can't return promise because it's already returns virtual DOM.

To repeat: react devs are agree that they made a mistake with this architecture but don't see any non-breaking approach to fix it. I read a lot of github threads / mail-lists on this topic, but have no time to google them again. Sorry about that. 

How it had to be done? Basically `this.state` should represent **pending** state. Then, asyncronicity is vanished from app space and remains only in implicit rendering layer. If you think about it, is sounds very logical. You don't care about the precise timestamp of component renderer. Moment later? No problem. But you do care about data consistency. About data flow in business processes. 
And async is the best way to trouble that.

Baobab behaves in the same "broken" way of React:

```javascript
let state = new Baobab({});
state.set("x", 1);
console.log(state.get("x")); // logs undefined
// visible state and message broadcasting is bound! 
```

How I imagine it should be:

```javascript
let state = new Baobab({});
state.set("x", 1);
console.log(state.get("x")); // logs 1
// visible state is pending state. Event broadcasting occurs somewhen later
// Now we have sync writes and reads!
```

I had not time to inspect how exactly Baobab is built but I think at least I exlpained my point of view in an understandable way. This way we don't need (at least) async vs sync option and chaining stuff.

**Comment by Yomguithereal** [comment]
Yes, your point is clear. Before I think about it more precisely, know you actually can force sync a single write operation with `tree.commit`.

Do you propose here to be able to write and read to the tree synchronously while notifying updates asynchronously?

**Comment by ivan-kleshnin** [comment]
> Before I think about it more precisely, know you actually can force sync a single write operation with tree.commit

Yes, it works. Thank you for the tip!

> Do you propose here to be able to write and read to the tree synchronously while notifying updates asynchronously?

Yes, I believe it should be better in a lot of aspects. 
But as there is a change I may be wrong, you better verify this theory on your own.

**Comment by Yomguithereal** [comment]
I'll propose a wa

**Comment by christianalfoni** [comment]
Yeah, this was my only concern, that the update has to be fired async or else we get a lot of unnecessary renders in components. 

I agree that there are not really any benefits to async commits, unless there is something "under the hood" that gets some performance boost? Nevertheless I can not imagine Baobab being the performance problem in any application. So a predictable synchronous behaviour is still worth it... as long as update events are async :-)

**Comment by ivan-kleshnin** [comment]
Great. I've got some experience trying to building an app on Baobab and I had to use `state.commit` just about everywhere. Every independent function / module must commit it's changeset otherwise dataflow is ruined. The only alternative is to make the whole tree synchronous with an init flag. But then I lose performance and step into rerendering madness. 

**1.1** should be huge. I mentioned that sync API should significantly improve client code.
But there are also a number of things that can be dropped from library after this: sync / async flags, commits, chains.

**Comment by Yomguithereal** [comment]
The  bonus with the update merging (which happen of course only when you have 2 updates or more, else I don't do it for perf reasons), is to be able to walk the tree only once when updating while "altering" the nodes only once. Since Baobab is a persistent data structure and some shallow copy is involved when walking the tree, this should represent a performance boost on large updates. But, to be honest, I never had time to "quantify" this perf boost.

It's right, however, that the merging part represent a large part of the application code and complexity and it would drastically simplify the library if we were to drop it. However I have a nasty intuition there is a trap here we are not seeing .

**Comment by ivan-kleshnin** [comment]
@Yomguithereal I see one potential pitfall when the data comes faster than applied. This is well known issue with streams and it can be solved by raising memory leak exception in our case. If we aren't sure about consequences, let's just make a branch and experiment there. I don't have much time to help with development but I can help with testing at least.

**Comment by christianalfoni** [comment]
Yeah, we have a pretty large application we can test it on. I can not really see that it could cause an issue as Baobab will conceptually just run multiple "tree merges" before the update, instead of one batched. There are no other side-effects on a merge than "update event"? 

Maybe stuff related to history etc.?

**Comment by Yomguithereal** [comment]
History works on events, so no problem here. I'll try to fiddle something in a branch for you to test soon.

**Comment by christianalfoni** [comment]
:+1:

**Comment by Yomguithereal** [comment]
@ivan-kleshnin: can you point me toward examples of your code where the asynchronous updates are an issue so I can assess the problem better please?

**Comment by ivan-kleshnin** [comment]
@Yomguithereal, sure. I'm going to do a cleanup and drop a link here.

**Comment by ivan-kleshnin** [comment]
@Yomguithereal you may inspect this repo:
https://github.com/Paqmind/react-starter/tree/master/

You should note a lot of `state.commit` calls, for example:
https://github.com/Paqmind/react-starter/blob/master/frontend/scripts/app.js#L55
https://github.com/Paqmind/react-starter/blob/master/frontend/scripts/robot/actions/set-offset.js#L10-L11

Without them it's impossible to read data in the deeper layer of data flow sequence.

**Comment by christianalfoni** [comment]
Hi guys, been thinking a bit more. If it goes synchronous it would be nice to have a `batch` method.

```js
tree.batch(function () {
  tree.set()...
});
```
That way if you have many operations you can still run them asynchronously. But I suppose you can not clean up the lib as much then? Anyways, just a thought :-)

**Comment by Yomguithereal** [comment]
You still have `tree.update` or `cursor.update`.

**Comment by Yomguithereal** [comment]
@ivan-kleshnin, @christianalfoni, I drafted a Baobab version that lets you write synchronously but fires its updates asynchronously. You can install it with `npm install baobab@1.1.0-syncwrite`. Could you try it and see whether this changes things?

**Comment by christianalfoni** [comment]
Cool, will try tomorrow!

Actually, I did get into trouble with async updates on `baobab-hot-loader`. Based on some diffs it needs to run multiple updates. That problem happened when running `apply` on a cursor after a `set`. The first set did not change or update? Sorry for bad info, my point was that there seems to be some edge case where async has problems.

**Comment by ivan-kleshnin** [comment]
> Could you try it and see whether this changes things?

Heck, yeah :smiling_imp:

**Comment by cassiozen** [comment]
I was also experiencing lost subsequent writes - this branch solved the issue for me.

**Comment by Yomguithereal** [comment]
@christianalfoni: in the current paradigm, running set, then apply in the same as running twice set on the same node and is consequently merged. One would have to use chain rather than apply to compose.

**Comment by ivan-kleshnin** [comment]
@Yomguithereal, so far so good :thumbsup:

**Comment by Yomguithereal** [comment]
What do you mean by charts @ivan-kleshnin? If you have some benchmarks, I'd love to see them.

> Well, as I understand repo is in flux state

What do you mean by this?

**Comment by ivan-kleshnin** [comment]
> What do you mean by this?

I said repo, but I  actually meant that **branch** contains a lot of TODO and excessive code because your aim was to get it fast as soon as possible, not to clean and optimize.

**Comment by Yomguithereal** [comment]
Synchronous write is now available on `master` through the `syncwrite` option (will default to `false`).

**Comment by ivan-kleshnin** [comment]
Great :+1:

**Comment by Yomguithereal** [comment]
Any issues with the `syncwrite` behavior so far? By the way, if you found a better name for the option than `syncwrite`...

**Comment by ivan-kleshnin** [comment]
> Any issues with the syncwrite behavior so far?
No, everything's fine with my usecases. Of course I'd like to hear more opinions.
Seems not so many people are using Baobab despite all the stars :smirk_cat:

**Comment by christianalfoni** [comment]
Hi guys, sorry for late response, I am "off the grid" working on a project at the moment :-)

Testing this out now!

**Comment by christianalfoni** [comment]
Hm, okay, it did not work "right out of the box". Got something not updating for some reason. Will check it out. Its a pretty big application so might be problematic to find the cause, but will give it a go :-)

For now I would say it is not safe to make `syncwrite` default at least. Keep you posted!

**Comment by christianalfoni** [comment]
Ah, I found the problem. If the mutation of the tree runs inside an "update" callback it does not trigger a new update. This is not good practice really, but yeah, that was that issue. Other than that it works great :-)

**Comment by Yomguithereal** [comment]
Hello @christianalfoni, could you post a link to the code you are describing or paste it somewhere here so we can check that?

**Comment by christianalfoni** [comment]
Jup, sure. The application code is way more complex than this, but this should be the core of the "problem". In principle it is not a good idea to run a new mutation on the "update" event of another as you could quickly cause some circular thing. I did not do the implementation of this logic on our project, but not sure if we could solve it any other way. Anyways, we fixed this issue by just wrapping the mutation in a setTimeout. Which I think is okay as it already is "special".

```js
var state = new Baobab({
  foo: 'bar',
  bar: 'foo'
});
```

state.select('foo').on('update', function () {
  state.set('bar', Date.now() + '-foo');
});

state.select('bar').on('update', function () {
  // Does not trigger when above callback runs "set"
});

state.set('foo', 'newValue');

**Comment by Yomguithereal** [comment]
I seems more like there is some kind of bug with the `syncwrite` behavior rather than it being expected to work this way. Your second callback should fire. It is probably because I emit the update event before resetting the tree's transaction or something of the kind.

**Comment by christianalfoni** [comment]
IMO it should just be ignored as it is bad practice. Mutating tree on tree update... skeptic ;-)

**Comment by ivan-kleshnin** [comment]
Maybe this is connected with https://github.com/Yomguithereal/baobab/issues/239

**Comment by Yomguithereal** [comment]
@christianalfoni, even if this is bad practice, there is probably a bug I should fix here :-).

**Comment by Yomguithereal** [comment]
Created issue #247.

**Comment by ivan-kleshnin** [comment]
Let's close older issues to not get overwhelmed.
